### PR TITLE
Auth Request Validation

### LINKED ISSUE #13: Authentication Request Validation
The Authorization Server MUST validate the request received as follows: 

1. The Authorization Server MUST validate all the OAuth 2.0 parameters according to the OAuth 2.0 specification.
2. Verify that a scope parameter is present and contains the openid scope value. (If no openid scope value is present, the request may still be a valid OAuth 2.0 request, but is not an OpenID Connect request.)
3. The Authorization Server MUST verify that all the REQUIRED parameters are present and their usage conforms to this specification.
4. If the sub (subject) Claim is requested with a specific value for the ID Token, the Authorization Server MUST only send a positive response if the End-User identified by that sub value has an active session with the Authorization Server or has been Authenticated as a result of the request. The Authorization Server MUST NOT reply with an ID Token or Access Token for a different user, even if they have an active session with the Authorization Server. Such a request can be made either using an id_token_hint parameter or by requesting a specific Claim Value as described in [Section 5.5.1](http://openid.net/specs/openid-connect-core-1_0.html#IndividualClaimsRequests), if the claims parameter is supported by the implementation.

As specified in OAuth 2.0 [RFC6749](http://tools.ietf.org/html/rfc6749), Authorization Servers SHOULD ignore unrecognized request parameters. 

If the Authorization Server encounters any error, it MUST return an error response, per [Section 3.1.2.6.](http://openid.net/specs/openid-connect-core-1_0.html#AuthError)

### PR DESCRIPTION
**Pull request by benfalk**
Validates an authenication request with some very simple checks
that are sure to grow as this project grows.  For now, this is
to only set the stage for the flow of what happens when the
request is invalid.

It should also be noted that currently it is returning a `400`
status code page with the errors; however, the the spec requires
that the user be redirected back to the request uri with added
parameters detailing the error:

http://openid.net/specs/openid-connect-core-1_0.html#AuthError

This will be written up as an additional feature to be completed
at a later date as more of the validation requirements are
understood.

#closes #13

### CODE REVIEW

File: web/models/authentication_request.ex [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    ui_locales: nil,
+    id_token_hint: nil,
+    login_hint: nil,
+    acr_values: nil,
+    flow_type: :invalid,
+    errors: [],
+    valid?: false
+  ]
+
+  @aloud_keys Keyword.keys(@structure) -- [:flow_type, :errors, :valid?]
```
Comment (scrogson) [inline web/models/authentication_request.ex]: `allowed_keys` 😉

Comment (benfalk) [inline web/models/authentication_request.ex]: my engrish isn't so good, thanks :+1:

File: web/models/authentication_request.ex [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    %{ request | errors: [] }
+    |> validate_scope
+    |> validate_key_present([:response_type, :client_id, :redirect_uri])
+    |> set_valid_flag
+  end
+
+  # ---------------------------[ Private Functions ]----------------------------
+
+  @doc false
+  defp extract_aloud_params(params) do
```
Comment (scrogson) [inline web/models/authentication_request.ex]: `extract_allowed_params`

File: web/models/authentication_request.ex [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  def validate(%__MODULE__{} = request, %Plug.Conn{} = _conn) do
+    %{ request | errors: [] }
+    |> validate_scope
+    |> validate_key_present([:response_type, :client_id, :redirect_uri])
+    |> set_valid_flag
+  end
+
+  # ---------------------------[ Private Functions ]----------------------------
+
+  @doc false
```
Comment (scrogson) [inline web/models/authentication_request.ex]: You don't need this for private functions

File: web/models/authentication_request.ex [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+     iex> %Ganondorf.AuthenticationRequest{response_type: "id_token token code"}
+     ...> |> Ganondorf.AuthenticationRequest.flow_type
+     :hybrid_flow
+
+     iex> %Ganondorf.AuthenticationRequest{response_type: "foobar"}
+     ...> |> Ganondorf.AuthenticationRequest.flow_type
+     :invalid
+  """
+  @spec flow_type(__MODULE__.t) :: __MODULE__.flow_type | :invalid
```
Comment (scrogson) [inline web/models/authentication_request.ex]: Can just use `t` instead of `__MODULE__.t`

Comment (scrogson) [inline web/models/authentication_request.ex]: Same with `flow_type` since these are local types

**Pull request merged**